### PR TITLE
#171380732 postgres db configurations

### PR DESCRIPTION
**Pull request by Muhire-Josue**
#### What does this PR do?

- It connects this project a Postgres database

#### Description of Task to be completed?

- Created a file in src/database/models that contains the connection to the database

- Includes database connection URI in .env file

- Created a script in package.json "connect:db" to test the connection

- If the connection was successful you get a message in the terminal, if the connection failed you get a message in the terminal

#### How should this be manually tested?

- git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend.git

- cd stackup2-barefoot-backend/

- npm install

- git fetch origin ch-postgres-configurations-#171380732

- git checkout ch-postgres-configurations-#171380732

- include the database URI in .env file, you can use .env.example for reference

- npm run db:connect

- npm start

#### Any background context you want to provide?

- This PR is to establish a connection between the application and a Postgres database

#### What are the relevant pivotal tracker stories?

- [171380732](https://www.pivotaltracker.com/story/show/171380732)

#### Screenshots (if appropriate)

- N/A

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,27 +4,36 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "start": "SET \"NODE_ENV=dev\" && nodemon ./src/index.js --exec babel-node --presets @babel/preset-env",
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "db:connect": "SET \"NODE_ENV=dev\" && babel-node ./src/database/models/ connect"
```
Comment (the22mastermind) [inline package.json]: Hi Josue! Can you make sure the command to connect to the db specified in the scripts matches the one in steps to test this PR? Keep up the good work.

File: src/index.js [hunk 2]
```diff
@@ -25,69 +17,9 @@ app.use(bodyParser.json());
...
-        }
-    });
-});
 
 // finally, let's start our server...
-const server = app.listen(process.env.PORT || 3000, function() {
+const server = app.listen(process.env.PORT || 3000, () => {
     console.log("Listening on port " + server.address().port);
 });
+export default app;
```
Comment (ricahcyuzuzo) [inline src/index.js]: You should need to add a new line at the end of your file?

File: src/routes/index.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
-const router = require("express").Router();
+import express from 'express';
 
+const router = express.Router();
 router.use("/api", require("./api"));
```
Comment (ricahcyuzuzo) [inline src/routes/index.js]: What does this means?

Comment (Muhire-Josue) [reply src/routes/index.js]: actually that is not my code.

**Review by muchristian — APPROVED** [review]
remember to add semicolon as how hound requesting, you did good job

**Review by ricahcyuzuzo — APPROVED** [review]
For me, It looks good!

File: .eslintrc.json [hunk 4]
```diff
@@ -9,6 +9,7 @@
   "rules": {
     "one-var": 0,
     "one-var-declaration-per-line": 0,
+    "linebreak-style": ["error", "windows"],
```
Comment (since9629) [inline .eslintrc.json]: why did you add this rule?

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
I personally appreciate the work you have done, you tried a lot and you did a really great job, congratulations, but I have two concerns; one is a suggestion another is a question.

### A.Suggestion
Could you add more info on how should you PR be manually tested ? 
=>You already have some but it seems to be insufficient
* git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend.git

* cd stackup2-barefoot-backend/

* npm install

* npm run db:connect

On these steps you specified, when you clone the repository, what you get on your local PC is the default branch `develop` this means your branch is not yet available, so you should add a command to run so that after cloning the repo, I'd be able to get your branch too, so as a suggestion you can add these two steps 
=>after that command of `cd stackup2-barefoot-backend/`
* `git fetch origin ch-postgres-configurations-#171380732`
* `git checkout ch-postgres-configurations-#171380732`

=>the continue with other remaining commands

### B.Question
When I manually tested your PR, I got the error you see on the screenshoot, can you help me to fix that?
It is telling me that `SET` not found,
i am using `Linux mint 19 Tessa`
Screenshot from 2020-02-28 14-52-45

File: package.json [hunk 5]
```diff
@@ -23,25 +25,25 @@
...
+    "mocha": "^7.1.0"
   },
   "devDependencies": {
     "eslint": "^6.1.0",
     "eslint-config-airbnb": "^18.0.1",
-    "eslint-plugin-import": "^2.18.2"
+    "eslint-plugin-import": "^2.18.2",
+    "morgan": "^1.9.1",
+    "nodemon": "^2.0.2",
+    "pg": "^7.18.2",
```
Comment (descholar-ceo) [inline package.json]: pg should be in dependecies not dev dependecies

File: package.json [hunk 6]
```diff
@@ -5,7 +5,9 @@
   "main": "index.js",
   "scripts": {
     "test": "nyc mocha --require @babel/register ./tests/*.js  --timeout 25000 --exit",
-    "coverage": "nyc report --reporter=text-lcov | coveralls"
+    "coverage": "nyc report --reporter=text-lcov | coveralls",
+    "start": "SET \"NODE_ENV=dev\" && nodemon ./src/index.js --exec babel-node --presets @babel/preset-env",
+    "db:connect": "SET \"NODE_ENV=dev\" && babel-node ./src/database/models/ connect"
```
Comment (descholar-ceo) [inline package.json]: for linux users, these `SET` gives error

**Review by descholar-ceo — APPROVED** [review]

File: .env.example [hunk 7]
```diff
@@ -0,0 +1,2 @@
+DATABASE_URL_DEV = postgres://{USER_NAME}:{USER_PASSWORD}@localhost:5432/{YOUR_DATABASE_NAME_DEV}
```
Comment (descholar-ceo) [inline .env.example]: this `.env.example` didn't work for me

**Review by Ugizwenayo-Divine — COMMENTED** [review]
You did great job but could you please put a space in start script in package.json
"start": "set NODE_ENV=dev&& nodemon ./src/index.js --exec babel-node --presets @babel/preset-env"
between dev and AND operator

**Review by the22mastermind — CHANGES_REQUESTED** [review]
I followed all the steps listed above but I am still unable to connect to the database. Can you try the above steps on your end and see if it works? Could you also remove the spaces around the "=" in your `.env` file? For example your write it like this: `VARIABLE_NAME=variablevalue`

**Review by the22mastermind — APPROVED** [review]
Looks good to me. Well done. Keep it up.

**Review by Ugizwenayo-Divine — APPROVED** [review]

**Comment by kipropbrian** [comment]
Check your commit message

File: package.json [hunk 8]
```diff
@@ -5,7 +5,9 @@
   "main": "index.js",
   "scripts": {
     "test": "nyc mocha --require @babel/register ./tests/*.js  --timeout 25000 --exit",
-    "coverage": "nyc report --reporter=text-lcov | coveralls"
+    "coverage": "nyc report --reporter=text-lcov | coveralls",
+    "start": "set NODE_ENV=dev&& nodemon ./src/index.js --exec babel-node --presets @babel/preset-env",
+    "db:connect": "set NODE_ENV=dev&& babel-node ./src/database/models/ connect"
```
Comment (kipropbrian) [inline package.json]: No need to set the node_env here. Lets make the env depend on what is set in the .env file. @karamuka please check this as it is related to your .env PR.

File: src/routes/api/users.js [hunk 9]
```diff
@@ -1,87 +0,0 @@
...
-const User = mongoose.model("User");
-
-router.get("/user", function(req, res, next) {
-    User.findById(req.payload.id)
-        .then(function(user) {
-            if (!user) {
-                return res.sendStatus(401);
-            }
-            return res.json({ user: user.toAuthJSON() });
-        })
```
Comment (kipropbrian) [inline src/routes/api/users.js:13]: Why is this file deleted?

Comment (Muhire-Josue) [reply src/routes/api/users.js:13]: Because part of my task was to get rid of mongoose in the project. and this particular file was implemented using mongoose and passport, that's why I erased its content.

File: .env.example [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#Development database
+DEV_DB_USERNAME=YOUR_USERNAME
+DEV_DB_PASSWORD=YOUR_PASSWORD
+DEV_DB_NAME=DATABASE_NAME
+DEV_DB_HOSTNAME=YOUR_HOSTNAME #ex: localhost
+DEV_DB_PORT=DATABASE_PORT_NUMBER #ex: 5432
```
Comment (descholar-ceo) [inline .env.example]: could you fix this newline issue?

File: src/database/models/index.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable no-unused-vars */
+import pg from 'pg';
+import dotenv from 'dotenv';
+
+const DATABASE_URI = `postgres://${process.env.DEV_DB_USERNAME}:${process.env.DEV_DB_PASSWORD}@${process.env.DEV_DB_HOSTNAME}:${process.env.DEV_DB_PORT}/${process.env.DEV_DB_NAME}`;
+dotenv.config();
```
Comment (descholar-ceo) [inline src/database/models/index.js:4]: This line of code `dotenv.config();` is supposed to come before line 5, an explanation is that: You can't access `process.env` before calling `dotenv.config()`

**Review by since9629 — CHANGES_REQUESTED** [review]
Could you add these ESLint rules at the end of ESLint rules:`
    "linebreak-style": 0,
    "no-console": 0,
    "no-unused-vars": 0

**Review by descholar-ceo — COMMENTED** [review]
And remember to deal with that `hound` warnings, as your codes will be our code-base structure

File: src/database/models/index.js (hunk 11 above)
Comment (Muhire-Josue) [reply src/database/models/index.js:4]: Ohh yes, I see. thanks

File: .eslintrc.json [hunk 12]
```diff
@@ -45,4 +60,4 @@
     "ecmaVersion": 6,
     "sourceType": "module"
   }
-}
+}
```
Comment (descholar-ceo) [inline .eslintrc.json]: newline issue

**Review by descholar-ceo — APPROVED** [review]
LGTM

**Review by kipropbrian — APPROVED** [review]
